adb/common_cli.py

```diff
@@ -19,144 +19,142 @@
-             **device_kwargs):
-  """Starts a common CLI interface for this usb path and protocol."""
-  argv = argv[1:]
-
-  if len(argv) == 1 and argv[0] == 'devices' and list_callback is not None:
-    # To mimic 'adb devices' output like:
-    # ------------------------------
-    # List of devices attached
-    # 015DB7591102001A        device
-    # Or with --output_port_path:
-    # 015DB7591102001A        device        1,2
-    # ------------------------------
-    for device in list_callback():
-      if FLAGS.output_port_path:
-        print '%s\tdevice\t%s' % (
-            device.serial_number,
-            ','.join(str(port) for port in device.port_path))
+class _PortPathAction(argparse.Action):
+  def __call__(self, parser, namespace, values, option_string=None):
+    setattr(
+        namespace, self.dest,
+        [int(i) for i in values.replace('/', ',').split(',')])
+
+
+class PositionalArg(argparse.Action):
+  def __call__(self, parser, namespace, values, option_string=None):
+    namespace.positional.append(values)
+
+
+def GetDeviceArguments():
+  group = argparse.ArgumentParser('Device', add_help=False)
+  group.add_argument(
+      '--timeout_ms', default=10000, type=int, metavar='10000',
+      help='Timeout in milliseconds.')
+  group.add_argument(
+      '--port_path', action=_PortPathAction,
+      help='USB port path integers (eg 1,2 or 2,1,1)')
+  group.add_argument(
+      '-s', '--serial',
+      help='Device serial to look for (host:port or USB serial)')
+  return group
+
+
+def GetCommonArguments():
+  group = argparse.ArgumentParser('Common', add_help=False)
+  group.add_argument('--verbose', action='store_true', help='Enable logging')
+  return group
+
+
+def _DocToArgs(doc):
+  """Converts a docstring documenting arguments into a dict."""
+  offset = None
+  in_arg = False
+  out = {}
+  for l in doc.splitlines():
+    if l.strip() == 'Args:':
+      in_arg = True
+    elif in_arg:
+      if not l.strip():
+        break
+      if offset is None:
+        offset = len(l) - len(l.lstrip())
+      l = l[offset:]
+      if l[0] == ' ':
+        out[m.group(1)] += ' ' + l.lstrip()
       else:
-        print '%s\tdevice' % device.serial_number
-    return
+        m = re.match(r'^([a-z_]+): (.+)$', l.strip())
+        out[m.group(1)] = m.group(2)
+  return out
+
+
+def MakeSubparser(subparsers, parents, method, arguments=None):
+  """Returns an argparse subparser to create a 'subcommand' to adb."""
+  name = ('-'.join(re.split(r'([A-Z][a-z]+)', method.__name__)[1:-1:2])).lower()
+  help = method.__doc__.splitlines()[0]
+  subparser = subparsers.add_parser(
+      name=name, description=help, help=help.rstrip('.'), parents=parents)
+  subparser.set_defaults(method=method, positional=[])
+  argspec = inspect.getargspec(method)
+  offset = len(argspec.args) - len(argspec.defaults or []) - 1
+  positional = []
+  for i in xrange(1, len(argspec.args)):
+    if i > offset and argspec.defaults[i-offset-1] is None:
```

Comment:
Add some comments as to what this is doing. It seems to be looking for all positional and non-None-defaulted keyword args and calling them positional?